apps/interactive-calibration/calibPipeline.cpp

```diff
@@ -30,37 +31,60 @@ CalibPipeline::CalibPipeline(captureParameters params) :
-    {
-        mCapture.open(mCaptureParams.camID);
-        cv::Size maxRes = getCameraResolution();
-        cv::Size neededRes = mCaptureParams.cameraResolution;
-
-        if(maxRes.width < neededRes.width) {
-            double aR = (double)maxRes.width / maxRes.height;
-            mCapture.set(cv::CAP_PROP_FRAME_WIDTH, neededRes.width);
-            mCapture.set(cv::CAP_PROP_FRAME_HEIGHT, neededRes.width/aR);
-        }
-        else if(maxRes.height < neededRes.height) {
-            double aR = (double)maxRes.width / maxRes.height;
-            mCapture.set(cv::CAP_PROP_FRAME_HEIGHT, neededRes.height);
-            mCapture.set(cv::CAP_PROP_FRAME_WIDTH, neededRes.height*aR);
-        }
-        else {
-            mCapture.set(cv::CAP_PROP_FRAME_HEIGHT, neededRes.height);
-            mCapture.set(cv::CAP_PROP_FRAME_WIDTH, neededRes.width);
+    auto open_camera = [this] () {
+        if(mCaptureParams.source == Camera)
+        {
+            mCapture.open(mCaptureParams.camID);
+            cv::Size maxRes = getCameraResolution();
+            cv::Size neededRes = mCaptureParams.cameraResolution;
+
+            if(maxRes.width < neededRes.width) {
+                double aR = (double)maxRes.width / maxRes.height;
+                mCapture.set(cv::CAP_PROP_FRAME_WIDTH, neededRes.width);
+                mCapture.set(cv::CAP_PROP_FRAME_HEIGHT, neededRes.width/aR);
+            }
+            else if(maxRes.height < neededRes.height) {
+                double aR = (double)maxRes.width / maxRes.height;
+                mCapture.set(cv::CAP_PROP_FRAME_HEIGHT, neededRes.height);
+                mCapture.set(cv::CAP_PROP_FRAME_WIDTH, neededRes.height*aR);
+            }
+            else {
+                mCapture.set(cv::CAP_PROP_FRAME_HEIGHT, neededRes.height);
+                mCapture.set(cv::CAP_PROP_FRAME_WIDTH, neededRes.width);
+            }
+            mCapture.set(cv::CAP_PROP_AUTOFOCUS, 0);
         }
-        mCapture.set(cv::CAP_PROP_AUTOFOCUS, 0);
+        else if (mCaptureParams.source == File)
+            mCapture.open(mCaptureParams.videoFileName);
+    };
+
+    if(!mCapture.isOpened()) {
+        open_camera();
     }
-    else if (mCaptureParams.source == File && !mCapture.isOpened())
-        mCapture.open(mCaptureParams.videoFileName);
     mImageSize = cv::Size((int)mCapture.get(cv::CAP_PROP_FRAME_WIDTH), (int)mCapture.get(cv::CAP_PROP_FRAME_HEIGHT));
 
     if(!mCapture.isOpened())
         throw std::runtime_error("Unable to open video source");
 
     cv::Mat frame, processedFrame, resizedFrame;
-    while(mCapture.grab()) {
+    while (true) {
+        if (!mCapture.grab())
+        {
+            if (!mCaptureParams.forceReopen)
+            {
+                CV_LOG_ERROR(NULL, "VideoCapture error: could not grab the frame.");
+                break;
+            }
+
+            do
+            {
+                open_camera();
+            } while (!mCapture.isOpened() || !mCapture.grab());
```

Comment:
Instead of blank screen for infinite period of time I prefer to see messages like these:
- "Reopening camera (try #N) ..."

Reply:
Thank you, I added "trying to reopen" message only before the loop since thousands of messages/sec about which try it is aren't very informative in my opinion.

Reply:
Why is not informative? It indicates that process hang or not.

To avoid thousands of messages/sec delay could be used, e.g. `cv::waitKey()` (also it reduce unnecessary CPU load)

Reply:
Because there are no functions that can hang, I think.(correct me if I'm wrong). CPU load is not an issue because this application already uses all resources it can find. Artificial delay would only slow us down if errors are common enough, but pass quickly. But if the delay is too short - we'll get a lot of messages if the errors are rare, but last longer. How many ms do you have in mind for delay?